### PR TITLE
Optimise template code by parsing only new events pushed to DL

### PR DESCRIPTION
**Pull request by shikharsharma-zeotap**
Optimise the template code with the following set of changes:
1. Stop memoizing dataLayer into a JSON to hold past event metadata, rather parse only new events pushed to it everytime.
2. Inject script only once.
3. Function refactors

### CODE REVIEW

File: template.tpl [hunk 1]
```diff
@@ -453,173 +453,145 @@ const url = 'https://content.zeotap.com/sdk/zeotap.min.js';
...
+  setInWindow('zeotap.callMethod', function () {
+    callInWindow('zeotap._q.push', arguments);
+  });
+  log('zeotap.callMethod', 'init', data.writeKey, options);
+  callInWindow('zeotap.callMethod', 'init', data.writeKey, options);
+  injectScript(url, data.gtmOnSuccess, data.gtmOnFailure, 'zeoCollect');
+}
 
-injectScript(url, data.gtmOnSuccess, data.gtmOnFailure, 'zeoCollect');
+if (!!dataLayer) {
```
Comment (satkuppu) [inline template.tpl]: Do we need to check for dataLayer array length as well?

File: template.tpl [hunk 2]
```diff
@@ -453,173 +453,145 @@ const url = 'https://content.zeotap.com/sdk/zeotap.min.js';
...
-  var k=0;
-  array.forEach((value)=>{
-    if(searchFn(value)){k=1;}
-  });
-  return k==1;
-}
-function isNamePresentIn(array,searchKey){
-  return searchWith(array,(value)=>searchKey == value.name);
+function isNamePresentIn(array, searchKey) {
+  for (let i = 0; i < array.length; i++) {
```
Comment (satkuppu) [inline template.tpl]: Can we use some prototype method?

File: template.tpl [hunk 3]
```diff
@@ -453,173 +453,145 @@ const url = 'https://content.zeotap.com/sdk/zeotap.min.js';
...
-    }
-    return mergedData;
-  }
-function isNullOrEmpty(o){
-    return o===null || o===undefined;
-} 
-function isEmptyString(s){
-    return isNullOrEmpty(s) || (typeof s === 'string' && s.length<=0);
+function isNullOrEmpty(o) {
+  return o === null || o === undefined;
```
Comment (satkuppu) [inline template.tpl]: can this be simplified with o == undefined?

File: template.tpl [hunk 4]
```diff
@@ -453,173 +453,145 @@ const url = 'https://content.zeotap.com/sdk/zeotap.min.js';
...
+}
+function isEmptyString(s) {
+  return isNullOrEmpty(s) || (typeof s === 'string' && s.length <= 0);
 }
-function getUserpropertiesFromData(eventData,propertiesList){
+function getUserpropertiesFromData(eventData, propertiesList) {
   var result = {};
-  for(let p = propertiesList.length-1;p>=0;p--){
+  for (let p = propertiesList.length - 1; p >= 0; p--) {
```
Comment (satkuppu) [inline template.tpl]: Why are we going backwards? Using Array reduce will make this function more readable

**Review by satkuppu — CHANGES_REQUESTED** [review]

File: template.tpl (hunk 2 above)
Comment (shikharsharma-zeotap) [reply template.tpl]: noted

File: template.tpl (hunk 1 above)
Comment (shikharsharma-zeotap) [reply template.tpl]: noted

File: template.tpl (hunk 3 above)
Comment (shikharsharma-zeotap) [reply template.tpl]: removed fn, since not used

File: template.tpl (hunk 4 above)
Comment (shikharsharma-zeotap) [reply template.tpl]: noted

**Pull request closed without merging**